fink_filters/ztf/filter_early_tde_candidates/lcs.py

```diff
@@ -0,0 +1,409 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+from astropy.table import Table, vstack
+from astropy.time import Time
+from astropy import units as u
+from astropy.coordinates import SkyCoord, concatenate
+
+import io
+import requests
+import pandas as pd
+
+from dustmaps.sfd import SFDQuery
+
+from light_curve.light_curve_py import RainbowFit
+
+import matplotlib.pyplot as plt
+
+import logging
+
+from .prefilter import mag2fluxcal
+
+COLORS_ZTF = {1: "#15284F", 2: "#F5622E"}
+
+API_ENDPOINT = 'https://api.fink-portal.org/api/v1/objects'
+
+# Filters ZTF
+filt_conv = {1: "g", 2: "r", 3: "i"}  # Conversion between filter ID (int) and filter name (str)
+band_wave_aa = {"g": 4770.0, "r": 6231.0, "i": 7625.0}  # Bands in angstroms for rainbow
+
+_LOG = logging.Logger("Early TDE")
+
+
+# TNS
+tns = None # Global cache for TNS entries
+
+def get_tns_info(oid=None, ra=None, dec=None, sr=5/3600, return_types=False):
+    """Return TNS classification for the object, if any"""
+    global tns
+
+    if tns is None:
+        r = requests.post(
+            'https://api.fink-portal.org/api/v1/resolver',
+            json={
+                'resolver': 'tns',
+                'name': '',
+                'nmax': 1000000
+            }
+        )
+
+        tns = pd.read_json(io.BytesIO(r.content))
+
+    if oid is not None:
+        # Search by ZTF objectId
+        idx = tns['d:internalname'] == oid
+    elif ra is not None and dec is not None:
+        # Search by coordinates
+        idx = SkyCoord(
+            tns['d:ra'].values,
+            tns['d:declination'].values,
+            unit='deg'
+        ).separation(
+            SkyCoord(ra, dec, unit='deg')
+        ).deg < sr
+    else:
+        return None
+
+    if return_types:
+        idx &= tns['d:type'] != 'nan'
+        return list(tns[idx]['d:type'])
+
+    result = []
+
+    for i,row in tns[idx].iterrows():
+        res = row['d:fullname']
+        if row['d:type'] and row['d:type'] != 'nan':
+            res += ' (' + row['d:type']
+            if row['d:redshift'] > 0:
+                res += f" z={row['d:redshift']}"
+            res += ')'
+
+        if res not in result:
+            result.append(res)
+
+    return " | ".join(result)
+
+
+# Extinction
+sfd = None
+# Coefficients for ZTF filters for G23 extinction law from dust_extinction for Rv=3.1
+Av = {1: 3.681, 2: 2.635, 3: 1.944}
+
+
+def deredden(flux, fid, ra, dec):
+    """De-reddening using SFD from dustmaps"""
+    global sfd
+
+    if sfd is None:
+        sfd = SFDQuery()
```

Comment:
To avoid any error in production in case the file is missing, I would rewrite it as:

```python
if sfd is None:
    try:
        sfd = SFDQuery()
    except FileNotFoundError as e:
        _LOG.warning("Dustmaps has not been initialized: {}".format(e))
        _LOG.warning("Downloading files now...")
        import dustmaps.sfd
        dustmaps.sfd.fetch()
        sfd = SFDQuery()
```

Reply:
I did it a bit differently, in a bit cleaner way code-wise - but probably it is a bit less error-prone if the download somehow fails mid-way. Anway, it is one-time process so maybe it would go?..